### PR TITLE
add a certificate signing request resource printer in kubectl

### PR DESCRIPTION
**Pull request by mikedanese**
#30163

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30166)

### CODE REVIEW

**Comment by luxas** [comment]
nice!

**Comment by pwittrock** [comment]
@mikedanese consider adding describe support as well

**Comment by mikedanese** [comment]
@pwittrock added descirbe

File: pkg/kubectl/describe.go [hunk 1]
```diff
@@ -1882,6 +1886,59 @@ func (p *PetSetDescriber) Describe(namespace, name string, describerSettings Des
...
+		return "", fmt.Errorf("Error parsing CSR: %v", err)
+	}
+
+	return tabbedString(func(out io.Writer) error {
+		fmt.Fprintf(out, "Name:\t%s\n", csr.Name)
+		fmt.Fprintf(out, "Labels:\t%s\n", labels.FormatLabels(csr.Labels))
+		fmt.Fprintf(out, "Annotations:\t%s\n", labels.FormatLabels(csr.Annotations))
+		fmt.Fprintf(out, "CreationTimestamp:\t%s\n", csr.CreationTimestamp.Time.Format(time.RFC1123Z))
+		fmt.Fprintf(out, "RequestingUser:\t%s\n", csr.Spec.Username)
+		fmt.Fprintf(out, "CommonName\t%s\n", cr.Subject.CommonName)
```
Comment (liggitt) [inline pkg/kubectl/describe.go]: should we include all the components of the subject? those could be significant if the approver issues a client certificate for this CSR (c.f. https://github.com/kubernetes/kubernetes/issues/30260)

File: pkg/kubectl/describe.go [hunk 2]
```diff
@@ -1882,6 +1886,70 @@ func (p *PetSetDescriber) Describe(namespace, name string, describerSettings Des
...
+		fmt.Fprintf(out, "Subject:\n")
+		fmt.Fprintf(out, "\tCommonName:\t%s\n", cr.Subject.CommonName)
+		fmt.Fprintf(out, "\tSerialNumber:\t%s\n", cr.Subject.SerialNumber)
+		fmt.Fprintf(out, "\tOrganization:\t%s\n", strings.Join(cr.Subject.Organization, ","))
+		fmt.Fprintf(out, "\tOrganizationalUnit:\t%s\n", strings.Join(cr.Subject.OrganizationalUnit, ","))
+		fmt.Fprintf(out, "\tCountry:\t%s\n", strings.Join(cr.Subject.Country, ","))
+		fmt.Fprintf(out, "\tLocality:\t%s\n", strings.Join(cr.Subject.Locality, ","))
+		fmt.Fprintf(out, "\tProvince:\t%s\n", strings.Join(cr.Subject.Province, ","))
+		fmt.Fprintf(out, "\tStreetAddress:\t%s\n", strings.Join(cr.Subject.StreetAddress, ","))
+		fmt.Fprintf(out, "\tPostalCode:\t%s\n", strings.Join(cr.Subject.PostalCode, ","))
```
Comment (liggitt) [inline pkg/kubectl/describe.go]: for the multi-value ones, can we print additional values on newlines to be able to distinguish between ["foo,bar"] and ["foo","bar"]? For example:
```
  Organization: foo
                bar
```
empty and single-value lists (the 99% case, mostly likely) wouldn't take up any more room

File: pkg/kubectl/describe.go [hunk 3]
```diff
@@ -1882,6 +1886,70 @@ func (p *PetSetDescriber) Describe(namespace, name string, describerSettings Des
...
+	if err != nil {
+		return "", fmt.Errorf("Error parsing CSR: %v", err)
+	}
+
+	return tabbedString(func(out io.Writer) error {
+		fmt.Fprintf(out, "Name:\t%s\n", csr.Name)
+		fmt.Fprintf(out, "Labels:\t%s\n", labels.FormatLabels(csr.Labels))
+		fmt.Fprintf(out, "Annotations:\t%s\n", labels.FormatLabels(csr.Annotations))
+		fmt.Fprintf(out, "CreationTimestamp:\t%s\n", csr.CreationTimestamp.Time.Format(time.RFC1123Z))
+		fmt.Fprintf(out, "RequestingUser:\t%s\n", csr.Spec.Username)
```
Comment (liggitt) [inline pkg/kubectl/describe.go]: I see other places in the describers using `"CreationTimestamp"`, but also plenty of places using title-case with spaces in labels. Can we do that here for all the multi-word ones (e.g. `Requesting User`, `Common Name`), except maybe `CreationTimestamp` to be consistent with the other uses of that.

File: pkg/kubectl/resource_printer.go [hunk 4]
```diff
@@ -1748,6 +1752,63 @@ func printClusterRoleBindingList(list *rbac.ClusterRoleBindingList, w io.Writer,
...
+	}
+	var approved, denied bool
+	for _, c := range csr.Status.Conditions {
+		switch c.Type {
+		case certificates.CertificateApproved:
+			approved = true
+		case certificates.CertificateDenied:
+			denied = true
+		default:
+			return fmt.Errorf("unkown csr conditon %q", c)
```
Comment (liggitt) [inline pkg/kubectl/resource_printer.go]: unknown

File: pkg/kubectl/describe.go [hunk 5]
```diff
@@ -1882,6 +1886,70 @@ func (p *PetSetDescriber) Describe(namespace, name string, describerSettings Des
...
+	csr, err := p.client.Certificates().CertificateSigningRequests().Get(name)
+	if err != nil {
+		return "", err
+	}
+
+	cr, err := utilcertificates.ParseCertificateRequestObject(csr)
+	if err != nil {
+		return "", fmt.Errorf("Error parsing CSR: %v", err)
+	}
+
```
Comment (liggitt) [inline pkg/kubectl/describe.go]: do we want an indication of the approval status (like in the printer), and whether or not a cert has been issued (nil check on status.cert)?

File: pkg/kubectl/resource_printer.go [hunk 6]
```diff
@@ -1748,6 +1752,63 @@ func printClusterRoleBindingList(list *rbac.ClusterRoleBindingList, w io.Writer,
...
+	}
+
+	if _, err := fmt.Fprintf(
+		w, "%s\t%s",
+		name,
+		translateTimestamp(meta.CreationTimestamp),
+	); err != nil {
+		return err
+	}
+	var approved, denied bool
```
Comment (liggitt) [inline pkg/kubectl/resource_printer.go]: do you think the requesting user and whether or not a cert has been issued should be in the printer, or should they have to switch to the describer to see that?

**Comment by liggitt** [comment]
a few nits, looks good otherwise

**Comment by mikedanese** [comment]
@liggitt PTAL

```console
$ kubectl get certificatesigningrequest     
NAME         AGE       REQUESTOR    CONDITION
mikedanese   2d        mikedanese   Pending
$ kubectl describe certificatesigningrequest
Name:                   mikedanese
Labels:                 <none>
Annotations:            <none>
CreationTimestamp:      Tue, 09 Aug 2016 10:46:09 -0700
Requesting User:        mikedanese
Status:                 Pending
Subject:
        Common Name:    mikedanese
        Serial Number:
Subject Alternative Names:
        DNS Names:              example.com
                                www.example.com
        Email Addresses:        foo@gmail.com
        IP Addresses:           10.10.10.10
No events.
```

**Comment by liggitt** [comment]
LGTM

**Comment by mikedanese** [comment]
Rebasing to fix compilation error cause by https://github.com/kubernetes/kubernetes/pull/28509

**Comment by mikedanese** [comment]
@k8s-bot ok to test, issue: #IGNORE

**Pull request merged**